pkg/controller/clusterdeployment/clusterinstalls.go

```diff
@@ -0,0 +1,239 @@
+		logger.WithError(err).Error("failed to get the cluster install")
+		return reconcile.Result{}, err
+	}
+
+	specModified := false
+	statusModified := false
+	// copy the cluster metadata
+	if met := ci.Spec.ClusterMetadata; met != nil {
+		if met.InfraID == "" ||
+			met.ClusterID == "" ||
+			met.AdminKubeconfigSecretRef.Name == "" ||
+			met.AdminPasswordSecretRef.Name == "" {
+			return reconcile.Result{}, nil // the information is not complete and therefore skip
+		}
+		if !reflect.DeepEqual(cd.Spec.ClusterMetadata, ci.Spec.ClusterMetadata) {
+			cd.Spec.ClusterMetadata = ci.Spec.ClusterMetadata
+			specModified = true
+		}
+	}
+
+	if cd.Status.InstallRestarts != ci.Status.InstallRestarts {
+		cd.Status.InstallRestarts = ci.Status.InstallRestarts
+		statusModified = true
+	}
+
+	// copy the required conditions
+	requiredConditions := []hivecontractsv1alpha1.ClusterInstallConditionType{
+		hivecontractsv1alpha1.ClusterInstallFailed,
+		hivecontractsv1alpha1.ClusterInstallCompleted,
+		hivecontractsv1alpha1.ClusterInstallRequirementsMet,
+	}
+
+	conditions := cd.Status.Conditions
+	for _, req := range requiredConditions {
+		cond := controllerutils.FindClusterInstallCondition(ci.Status.Conditions, req)
+		if cond == nil {
+			continue
+		}
+
+		updated := false
+		conditions, updated = controllerutils.SetClusterDeploymentConditionWithChangeCheck(conditions,
+			hivev1.ClusterDeploymentConditionType("ClusterInstall"+string(cond.Type)), // this transformation is part of the contract
+			cond.Status,
+			cond.Reason,
+			cond.Message,
+			controllerutils.UpdateConditionIfReasonOrMessageChange,
+		)
+		if updated {
+			statusModified = true
+		}
+	}
+	cd.Status.Conditions = conditions
+
+	// take actions based on the conditions
+	// like,
+	// update install started time when requirements met.
+	// update installed = true when completed
+	// update the installed timestamp when complete
+
+	requirementsMet := controllerutils.FindClusterDeploymentCondition(conditions, hivev1.ClusterInstallRequirementsMetClusterDeploymentCondition)
+	if requirementsMet != nil &&
+		requirementsMet.Status == corev1.ConditionTrue {
+		if !reflect.DeepEqual(cd.Status.InstallStartedTimestamp, &requirementsMet.LastTransitionTime) {
+			cd.Status.InstallStartedTimestamp = &requirementsMet.LastTransitionTime
+			statusModified = true
+
+			kickstartDuration := time.Since(ci.CreationTimestamp.Time)
+			logger.WithField("elapsed", kickstartDuration.Seconds()).Info("calculated time to first provision seconds")
+			metricInstallDelaySeconds.Observe(float64(kickstartDuration.Seconds()))
+		}
+	}
+
+	completed := controllerutils.FindClusterDeploymentCondition(conditions, hivev1.ClusterInstallCompletedClusterDeploymentCondition)
+	if completed != nil && completed.Status == corev1.ConditionTrue { // the cluster install is complete
+		cd.Spec.Installed = true
+		cd.Status.InstalledTimestamp = &completed.LastTransitionTime
+		specModified = true
+		statusModified = true
```

Comment:
Are we sure we modified here? It looks like installed and timestamp may already have been set and we're re-reconciling?

Reply:
the reconcile here is for "installing" clusterinstall , when we set the installed = true we are no longer reconciling.
so in sense this is one time update when we see Completed = true, so there is no need for compare and update